dialogue-apache-hc4-client/src/main/java/com/palantir/dialogue/hc4/ApacheClientGauges.java

```diff
@@ -35,28 +43,60 @@
 @SuppressWarnings("UnnecessaryLambda")
 final class ApacheClientGauges {
 
-    private static final List<WeakReference<PoolingHttpClientConnectionManager>> pools = new CopyOnWriteArrayList<>();
-    private static final Gauge<Long> idle =
-            () -> collect(pool -> (long) pool.getTotalStats().getAvailable(), Long::sum, 0L);
-    private static final Gauge<Long> leased =
-            () -> collect(pool -> (long) pool.getTotalStats().getLeased(), Long::sum, 0L);
-    private static final Gauge<Long> pending =
-            () -> collect(pool -> (long) pool.getTotalStats().getPending(), Long::sum, 0L);
+    private static final ConcurrentMap<String, List<WeakReference<PoolingHttpClientConnectionManager>>> poolsByName =
+            new ConcurrentHashMap<>();
+    private static final TaggedMetricSet metricSet = new TaggedMetricSet() {
+        @Override
+        public Map<MetricName, Metric> getMetrics() {
+            ImmutableMap.Builder<MetricName, Metric> results = ImmutableMap.builder();
+            forEachMetric(results::put);
+            return results.build();
+        }
+
+        @Override
+        public void forEachMetric(BiConsumer<MetricName, Metric> consumer) {
```

Comment:
does the specialized implementation of this really buy us anything? esp since we're already adding them all into a `new DefaultTaggedMetricRegistry()`, seems like we could just move this code into `getMetrics()` and then rely on the default impl?

Reply:
Adding the version when metrics are created allows us to do the work once when the metric is created/accessed as opposed to each time metrics are loaded, in theory. However quite a few places create metric names with service/endpoint information which moves that work to each request... so I think you're right, it's best if we tag on access.
Thoughts on making that change in this PR vs an isolated change?